src/DnsClient/DnsDatagramReader.cs

```diff
@@ -275,8 +279,13 @@ public DnsString ReadQuestionQueryString()
             return DnsString.FromResponseQueryString(value);
         }
 
-        public ICollection<ArraySegment<byte>> ReadLabels()
+        public ICollection<ArraySegment<byte>> ReadLabels(ref int recursion)
```

Comment:
why by ref? would be much easier to just use `int recursion = 0` so that no one has to initialize it and there is no code change elsewhere.

Also, more importantly, re initializing the counter below where we do the recursion basically makes this not do anything.